Neos.Flow/Classes/Mvc/Routing/RouteAnnotationRoutesProvider.php

```diff
@@ -0,0 +1,107 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Neos\Flow\Mvc\Routing;
+
+use Neos\Flow\Mvc\Exception\InvalidActionNameException;
+use Neos\Flow\Mvc\Routing\Exception\InvalidControllerException;
+use Neos\Flow\ObjectManagement\ObjectManagerInterface;
+use Neos\Flow\Reflection\ReflectionService;
+use Neos\Flow\Annotations as Flow;
+use Neos\Utility\Arrays;
+
+class RouteAnnotationRoutesProvider implements RoutesProviderWithOptionsInterface
+{
+    /**
+     * @param ReflectionService $reflectionService
+     * @param ObjectManagerInterface $objectManager
+     * @param array<string> $classNames
+     */
+    public function __construct(
+        public readonly ReflectionService $reflectionService,
+        public readonly ObjectManagerInterface $objectManager,
+        public readonly array $classNames = [],
+    ) {
+    }
+
+    /**
+     * @param array<string, mixed> $options
+     * @return $this
+     */
+    public function withOptions(array $options): static
```

Comment:
This is a pragmatic approach that works due to our crazy dependency injection that can deal with prototype injection. But it can lead to weird (caching) issues if the object is called before it is "optionized" and IMHO a factory pattern would be the more explicit way to solve this, i.e.:

```php
interface RoutesProviderFactoryInterface {
    public function create(array $options): RoutesProviderInterface
}

#[Flow\Scope('singleton')]
final class AttributeRoutesProviderFactory implements RoutesProviderFactoryInterface {
    public function __construct(
        public readonly ReflectionService $reflectionService,
        public readonly ObjectManagerInterface $objectManager,
    ) {}

    public function create(array $options): RouteAnnotationRoutesProvider {
        return new RouteAnnotationRoutesProvider($this->reflectionService, $this->objectManager, $options['classNames']);
    }
}

final class RouteAnnotationRoutesProvider implements RoutesProviderInterface {
    public function __construct(
        public readonly ReflectionService $reflectionService,
        public readonly ObjectManagerInterface $objectManager,
        public readonly array $classNames,
    ) {}
    
    // ...
}
```

as a consequence we should specify the *factory*  instead of the *provider*  class in the settings:

```yaml
providerFactory: Some\Namespace\AttributeRoutesProviderFactory
providerOptions:
  foo: bar
```

That might seem cumbersome but we use that pattern a lot in Neos 9 and IMO it works very well

Reply:
Will adjust that tomorrow ... that will avoid the RouteProviderWithOptions problem that is weird aswell ... i just wanted to avoid builderFactoryGenratorProviders

Reply:
That's great. And it will allow us to extend the feature just by tweaking the factory.
For example it could support an option 'packageKey' and translate it to class names without having to change the provider itself